### PR TITLE
Move commits of old PR #593 on assignment to a new one post-rebase

### PR DESCRIPTION
**Pull request by pchauchat**
This PR allows gs.assignment to work with booleans (and indices given as tensors) in tf.

It is a follow-up on #583 and #593 

gs.assignment was heavily refactored to take into account arrays and booleans. For numpy and pytorch, I tried to take as much advantage as possible of their slicing opportunities, to keep the cost low (although I don't guarantee anything).

I defined more general check functions, to see if an object is some kind of boolean (mostly a native one or an array of bools), and if it is iterable (list, tuple, array of dim > 0). Maybe it will be useful for other functions too.

Some lines are left untested, I will add the corresponding tests later.

### CODE REVIEW

File: geomstats/geometry/hypersphere.py [hunk 1]
```diff
@@ -300,12 +307,13 @@ def random_uniform(self, n_samples=1):
...
             norms = gs.linalg.norm(samples, axis=1)
-            indcs = gs.isclose(norms, 0.0)
+            indcs = gs.isclose(norms, 0.0, atol=tol)
             num_bad_samples = gs.sum(indcs)
             if num_bad_samples == 0:
                 break
-            samples[indcs, :] = gs.random.normal(
+            new_samples = gs.random.normal(
                 size=(num_bad_samples, self.dim + 1))
+            samples = self._replace_values(samples, new_samples, indcs)
```
Comment (nguigs) [inline geomstats/geometry/hypersphere.py:316]: I came up with an idea:
`samples = gs.concatenate([samples[~indcs], new_samples])`
what do you think?

Comment (pchauchat) [reply geomstats/geometry/hypersphere.py:316]: That's smart, since we don't care about the order !

However, to have that part of code tested, we either have to do "seed mining", or define it in an auxiliary function. Any idea of how it could be named ? (replace_values would not be completely appropriate anymore)

Comment (nguigs) [reply geomstats/geometry/hypersphere.py:316]: True, maybe `_enforce_non_zeros`

Comment (pchauchat) [reply geomstats/geometry/hypersphere.py:316]: It does not explictely deal with the zeros, simply the replacement. Maybe _replace_unordered_values ?
I don't know whether it could be interesting to have both (ordered and unordered alternatives).

File: geomstats/_backend/numpy/__init__.py [hunk 2]
```diff
@@ -118,6 +118,24 @@ def get_mask_i_float(i, n):
     return mask_i_float
 
 
+def _is_boolean(x):
+    if isinstance(x, bool):
+        return True
+    if isinstance(x, (tuple, list)):
+        return _is_boolean(x[0])
```
Comment (ninamiolane) [inline geomstats/_backend/numpy/__init__.py:125]: Why looking at only the first element of the tuple/list? It is assumed that all elements have the same type?

File: geomstats/_backend/tensorflow/__init__.py [hunk 3]
```diff
@@ -341,11 +365,19 @@ def _assignment_single_value(x, value, indices, axis=0):
         Copy of x where value was assigned at all indices (and possibly
         along an axis).
     """
+    if _is_boolean(indices):
+        indices = [index for index, val in enumerate(indices) if val]
```
Comment (ninamiolane) [inline geomstats/_backend/tensorflow/__init__.py:369]: This line comes up often, and could be factorize into a `_from_bools_to_indices` aux function.

File: geomstats/_backend/numpy/__init__.py (hunk 2 above)
Comment (ninamiolane) [reply geomstats/_backend/numpy/__init__.py:125]: Just saw your answer in the former PR:

I made the assumption that either all elements are booleans or no one is. I don't think it makes sense to make the code heavier and longer to run to be able to account for cases where booleans would be mixed with non-boolean values. I am not even sure Numpy handles them.

Comment (pchauchat) [reply geomstats/_backend/numpy/__init__.py:125]: I made the assumption that either all elements are booleans or no one is. I don't think it makes sense to make the code heavier and longer to run to be able to account for cases where booleans would be mixed with non-boolean values. I am not even sure Numpy handles them.

File: geomstats/_backend/tensorflow/__init__.py (hunk 3 above)
Comment (pchauchat) [reply geomstats/_backend/tensorflow/__init__.py:369]: True ! Do you think this should stay here, or could be useful elsewhere and go to the backend ? (In which case, a subsequent PR could be appropriate.)
